### ISSUE TITLE
How to publish different semver changes in independent mode in a CI

### ISSUE THREAD

**Issue by bdwain** [lerna/lerna#1286]
## Expected Behavior
Currently, the suggested way to run publish on the CI is to use the --cd-version and --yes flags. However, --cd-version only lets you specify a single change type (major/minor/patch/etc..) for all affected packages instead of allowing more granular control (for independent mode at least). As far as I can tell, no option exists to enable this on the command line. 

## Current Behavior
Ideally, there would be some way to say something like "package 1 gets a major bump, package2 gets a patch, package3 gets a minor, etc...", and maybe also a default for unspecified packages. This may be awkward to do on the command line manually, though since there's already a fine solution to this problem when not on a CI, I don't think that's too much of an issue since this command would most likely be auto-generated. 

## Possible Solution
the 2 simplest things would be to allow something like this
`lerna publish --versions "package1,major,package2,minor"`
or
`lerna publish --versions-file <path_to_file>`

Another option would be if there was a node api for lerna, it could be done with that. I haven't seen one though.

## Context
I'm trying to work on moving my project to a monorepo. We need to do our publish commands in our CI because they are hitting a private npm registry. We often have breaking changes in one package that are not breaking changes when pulled in to other packages. It would be good to avoid a major bump in one package forcing a major bump in all packages that depend on it. 

What I'm imagining I'd have to do now would be to ignore the publish command altogether and write my own  script that parsed the results of `lerna updated --json` and bumped each package manually using the semver i specify for that package, and then just running npm publish directly on the package.

**Comment by evocateur** [comment]
This sounds like a job for `--conventional-commits`.

**Comment by bdwain** [comment]
I saw that but I don't think conventional commits does what I'm describing. 

For example, lets say i have a component in package1, and package2 uses it. Let's say they're both v1.0.0 right now (but independent) and that I make a breaking change to the component where i rename an option. In my commit, i rename the option in package1, and update the usage of it in package2 to pass the new option name. 

I want to bump package1 to 2.0.0 and package2 to 1.0.1. I have a single commit, and I could put BREAKING CHANGE in it to signify a major bump, but that would bump both packages to 2.0.0. It seems like I could make it work if i avoided squashing and made sure to do 1 commit per package, but that is kind of an artificial restriction and not being able to squash would be annoying to deal with.

**Comment by evocateur** [comment]
Or just ...accept that it’s a breaking change? I mean, it’s pretty trivial to split the commits. I don’t see how you even configure the convoluted solution you’re proposing, since it would need to be different literally every time you pushed to CI.

**Comment by bdwain** [comment]
It's not a breaking change to package2 though. Only package 1. Imagine package2 is lodash and package1 is some helper that lodash used under the hood. If I am an app that uses lodash, I don't need to change my usages of lodash when upgrading from the old version to the new one, so why would it be a major bump? Lodash hides the usage of package1 from me, so making it a breaking change just because package1 had a breaking change would be wrong. The same logic could apply to any dependency otherwise, not just ones in the same project.

While it's not difficult to split the commits, probably the most important benefit of monorepo is to streamline changes that affect multiple packages at once, so artificially splitting a single change up into separate commits seems to somewhat defeat the purpose.

They are not necessarily the only solutions, but two that could work are the ones i described above.

```js
lerna publish --versions "package1,major,package2,minor"
or
lerna publish --versions-file <path_to_file>
```

it's a bit awkward to type that out manually, but they are there mainly for CI usage anyway, since the current version of publish that prompts you for each package works fine for a human.

**Comment by evocateur** [comment]
How exactly is this list of hints supposed to make its way into an automated CI process, though?

I’m increasingly convinced the “independent” versioning mode was a mistake. All it does is make people throw random packages into a single repository and then complain when they have to version them at the same time.

**Comment by bdwain** [comment]
The way I imagined this working (at least for my use case, which is not necessarily how it would be for everyone else) is that the jenkins job that would be responsible for bumping versions would have some sort of input where you can choose packages from the list of available packages in the repo and apply a semver to them. Any unchosen packages that are going to be published would get some default semver bump (probably a patch).

So in the example above, in my job i would specify that package1 would get a major bump, and lodash, being unspecified, would get a patch bump.

My lodash example was obviously contrived but I think there are plenty of valid use cases for independent mode. My project currently contains a few different apps and they all depend on a single common library, so a monorepo helped making changes to all of them at once. But some of the apps share other helper libraries that not everyone uses. I wouldn't want a major bump in one of those libraries to force me to bump every app, especially because i'd like to automate releases on changes and I don't want to release when nothing actually changed.

**Comment by morewry** [comment]
When I use `lerna publish --independent --conventional-commits --yes --message 'chore: release'`, I get what it sounds to me like you want in every case I've tested but one. That one is the one you want, but take a look. I'm pretty satisfied with the behavior, aside from this one case.

I feel compelled to say here. I would be very unhappy using Lerna without independent versioning, and I'm not "_throwing random packages into a single repository and then complaining when I have to version them at the same time_." I just want my tools to help me communicate accurately with semver instead of giving false alarms that a breaking change has occurred when actually an implementation detail changed.

Say I have 10 different modules in `packages/` without dependency relationships. I make commits on 3 of those modules: `a`, `b`, and `c`:
- In module `a` I make a `fix` commit.
- In module `b` I make a `feat` commit.
- In module `c` I make a `feat` commit with a `BREAKING CHANGE`.
- Modules `d` - `j` are unchanged.

When I publish, I get the following:
- Module `a` gets a new patch version.
- Module `b` gets a new minor version.
- Module `c` gets a new major version.
- Modules `d` - `j` are unchanged.

Say these modules do have dependency relationships:
- Module `d` depends on `a`.
- Module `e` depends on `b`.
- Module `f` depends on `c`.
- Modules `g` - `j` are loners.

When I publish with the same changes to `a`, `b`, and `c`, I get the following:
- Module `a` gets a new patch version.
- Module `b` gets a new minor version.
- Module `c` gets a new major version.
- Module `d` gets a new patch version with a new minimum version for `a`.
- Module `e` gets a new patch version with a new minimum version for `b`.
- Module `f` gets a new patch version with a new minimum version for `c`.
- Modules `g` - `j` are are unchanged.

As you mentioned, all this relies on making separate commits. This is the effect you'd get if you were developing directly on `master`. Which begs the question of pull requests and merges. There are two main cases:

- If all these changes are completely unrelated tasks that just happen to affect modules in a dependency relationship with each other, you would ideally open separate pull requests. You can rebase, squash, or regular merge. This case is mostly straightforward.

- If all of these changes were part of a related task, you would prefer to open a single pull request. This case is less straightforward. I get a different result with squash merges than with rebase or regular merges--but rebase and regular merges work as expected.

    - With a _rebase_ or _regular_ merge, module `f` gets a new patch version
    - With a _squash_ merge, module `f` can end up a new major version

It's debatable whether it should be "fixed" and I don't know how feasible it is, but I think there's a case to be made, although I'm too tired to make it at this point. Overall I'm just not totally happy with how squash commits work with publish, even though I prefer them for ease of reasoning about history. Alternative ways of handling them might have too many downsides; I don't have enough context.

I'll wrap up with how the squash case looks. In my tests, I intentionally didn't write my PR as a conventional commit. I just let it be the name of my branch and used Github's default commit message, which combines the commit messages of all of the squashed commits, like so:

```
commit xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Author: Xxx Xxxxx <xxxx@xxxxxx.xxx>
Date:   Xxx Xxx xx xx:xx:xx xxxx -xxxx

    Example/pullrequest (#xx)

    * feat(c): breaking change to c

    BREAKING CHANGE:
    changed something in c's API

    * feat(b): add feature to b

    * fix(a): fix bug in a
```

The `CHANGELOG` for `c`, `b`, and `a` all get this entry:

```
<a name="3.0.0"></a>
# 3.0.0 (xxxx-xx-xx)

* Example/pullrequest (#xx) ([xxxxxxx](https://github.com/xxxxxxx/xxxxxxx/commit/xxxxxxx))

### BREAKING CHANGES

* changed something in c's API

* feat(b): add feature to b

* fix(a): fix bug in a
```

I'm tempted to think that if a commit looks like this, it'd be nice to provide special support for Github's default format for the purposes of the `CHANGELOG`, putting aside the matter of how the changes _should_ be bumped. It'd just make it easier to if you were doing conventional commits along the way not to have to edit that default commit and still get a good `CHANGELOG`.

**Comment by evocateur** [comment]
As nice as that would be, it’s suuuuper out of scope for lerna directly. You’ll need to bring it up with the conventional-changelog folks, or possibly develop a custom preset.

**Comment by bdwain** [comment]
@evocateur are you open to the solution I suggested above or anything similar to solve the issue of applying different semvers to different packages via CI? I think that at least is definitely in scope for lerna since publish isn't really usable for a completely valid set of semver changes without it.

**Comment by evocateur** [comment]
I'm not too tickled about making `lerna publish` even more complicated, tbh. Improvements to `--conventional-commits` are welcome, though.

**Comment by bdwain** [comment]
I'll probably just write a custom publish script instead. I don't think conventional commits is the solution to this.

**Comment by fabioberger** [comment]
@evocateur Since `lerna publish` already asks the developer to select a version for each package in an interactive mode, why not also add a command-line way of doing the same thing?

I agree with @bdwain that conventional commits is only one way to handle this limitation and it requires adopting the conventional commit workflow, which many don't want to do.

**Comment by bdwain** [comment]
@evocateur what do you mean for "each ci run"? whether the command line params are `--cd-version major` or `--versions "package1,major,package2,minor,package3,patch"` it seems the same to me as far as per-run configuration.

**Comment by bdwain** [comment]
The idea was not to hard-code the string (`package1,major,package2,minor,package3,patch`), but have it calculated and passed from whatever is running `lerna publish`  (like a jenkins build script) based on params passed to the job. It's a little awkward to type as a human, but it's meant for CI anyway so who cares. Interactive mode is fine for humans.

**Comment by evocateur** [comment]
How does Jenkins determine that value? If it's not automated, that's per-run config that a human has to decide.

**Comment by bdwain** [comment]
Isn't `--cd-version major` also a per run config that a human has to decide? it's just simpler because there's only one version. That would be a single jenkins job param. The multi-version one would just take multiple params.

My thought was that you would specify all of the semver bumps you want to specify, and maybe a default bump for unspecified packages that have changed. So if i wanted to say package1 was major, pacakge2 was minor, and every other package that was affected was a patch, i would say

`lerna publish --semvers "package1,major,package2,minor" --default-semver "patch"`

And then in the jenkins job I would just have to somehow specify package1 and package2 and the default. Every CI and job is different though, so that's not really lerna's concern. They just need to pass the right params to publish based on the job params.

Patch could even be the default value for default-semver or something to avoid needing that every time.

**Comment by fabioberger** [comment]
@evocateur in our case we decide the next version for each package in a CHANGELOG file in each sub-package, and want to have a publish script that pulls the versions from there and calls `lerna publish` with the proper next versions. It isn't impossible to imagine that people might have a different way of keeping track and determining future versions that doesn't involve conventional commits.

**Comment by morewry** [comment]
I had a situation recently where I had a minor version bump (new feature) in a package (`package-a`). `package-a` is a dependency of another package (`package-b`), so it got bumped too, as a patch. But in this dependency relationship, the new feature in `package-a` leads directly to a new feature in `package-b`, so minor would be more appropriate. I use `--conventional-commits`, and I found myself wishing there was a CLI param that would let me tweak `package-b`'s version change. It got me thinking back to this issue and how this behavior in Conventional Commits directly stems from Lerna's orchestration because Lerna creates the commit in `package-b` that leads to the patch. AFAIK there isn't even an interactive workaround available, however, in a case like this. Figured I'd share the scenario and chime in saying that while I'm pretty satisfied with Lerna and `--conventional-commits`, I do think there's still room for improvement. The use cases are diverse and quirky, so it's not gonna be easy, but I think it's worth continuing to think about and experiment with.

**Comment by k0nserv** [comment]
We are also very interested in the ability to publish different versions in independant mode without using the interactive method. For us the burden of manually chosing versions is getting significant and prone to human error. We are experimenting with a [method](https://github.com/Skyscanner/backpack/pull/642/) that can determine what changes have occured in the package and thus what semver change needs to be made. Something like `lerna publish --force-publish package-1@1.3.2,package-2@2.3.4,package@3.2.1` or what @bdwain suggested would be very useful for us. We aren't really interested in using `--conventional-commits` as it doesn't work well with out workflow and process.

**Comment by fabioberger** [comment]
This is our work-around at the moment, in case others need this functionality immediately. We simply commandeer the interactive prompt programmatically. I would much prefer this be supported natively by the command-line interface.

```typescript
async function lernaPublishAsync(packageToVersionChange: { [name: string]: string }) {
    // HACK: Lerna publish does not provide a way to specify multiple package versions via
    // flags so instead we need to interact with their interactive prompt interface.
    const child = spawn('lerna', ['publish', '--registry=https://registry.npmjs.org/'], {
        cwd: constants.monorepoRootPath,
    });
    let shouldPrintOutput = false;
    child.stdout.on('data', (data: Buffer) => {
        const output = data.toString('utf8');
        if (shouldPrintOutput) {
            utils.log(output);
        }
        const isVersionPrompt = _.includes(output, 'Select a new version');
        if (isVersionPrompt) {
            const outputStripLeft = output.split('new version for ')[1];
            const packageName = outputStripLeft.split(' ')[0];
            let versionChange = packageToVersionChange[packageName];
            const isPrivatePackage = _.isUndefined(versionChange);
            if (isPrivatePackage) {
                versionChange = 'patch'; // Always patch updates to private packages.
            }
            const semVerIndex = semverNameToIndex[versionChange];
            child.stdin.write(`${semVerIndex}\n`);
        }
        const isFinalPrompt = _.includes(output, 'Are you sure you want to publish the above changes?');
        if (isFinalPrompt && !IS_DRY_RUN) {
            child.stdin.write(`y\n`);
            // After confirmations, we want to print the output to watch the `lerna publish` command
            shouldPrintOutput = true;
        } else if (isFinalPrompt && IS_DRY_RUN) {
            utils.log(
                `Submitted all versions to Lerna but since this is a dry run, did not confirm. You need to CTRL-C to exit.`,
            );
        }
    });
}
```

**Comment by bdwain** [comment]
reopening this as there's clearly a lot of interest besides just me. 

I'm not particularly attached to any solution, but it seems clear that this is a clear piece of functionality is something lerna should have to fully support independent mode. 

I think something like `lerna publish --semvers "package1,major,package2,minor" --default-semver "patch"` would address the issue and be relatively easy to implement. It sounds like there's plenty of volunteers to implement it as long as you're ok with the feature @evocateur. Thoughts?

**Comment by bdwain** [comment]
@evocateur any chance you could reconsider this. I think the comments on this issue demonstrate a clear area where lerna is currently lacking, and a clear solution to that problem. The only available solutions right now are super hacky like intercepting the interactive prompt programmatically or avoiding lerna publish altogether. Neither of those seem very reasonable.

I think there's lots of people who would be happy to make the PR as long as you're open to the idea.

**Comment by GordonSmith** [comment]
Question:  What happens when you have several branches:
* ver1.x (where 1 is the largest major version for all packages)
* ver2.x (where 2...)
* ver3.x (...)

Whenever a breaking change PR is made, it would need to be issued against the "next version up" branch.  Then in that branch your major versions will get bumped on publish, but in the other branch they won't be.

Now (in theory) you can continue to publish to NPM from any of the verX branches - but immediately afterwards you would want to upmerge  ver1 -> ver2 -> ver3 -> master before doing another publish (from a higher version).

I haven't tested the above, but I am also looking for a solution where I can bump a major version on a util package without unnecessarily bumping a major version on its dependents (which my compiler will tell and which will be included in the breaking PR).

**Comment by light24bulbs** [comment]
I agree, this feature is a necessary part of independent mode.  If the versions are independent, we naturally don't necessarily want to bump them all the same version, especially if all that has changed is a version number in the dependencies.  

I also think we should be able to do this manually rather than being forced to use a commit message DSL.

I understand that Lerna's core team is mostly centered around locked mode, and that's great, but lots of people are using Lerna these days for use cases that aren't babel-like.  

There are two ways I can imagine this working:  One is a manual command line flag as described above, and the other is a flag in the projects `package.json`.  

`"lerna-bump-type": "major"`

That lerna takes out when it changes the version.  That's because most sane people do Lerna publish from a CI build, and it would be nice to see that the breaking change is acknowledged in the PR before anyone clicks submit.  Anyway, if Lerna adds a manual flag, that is how I will probably write my pipeline.

**Comment by fabioberger** [comment]
This is still an existing issue for us. We are currently using a fork of Lerna to get around it but it would be great to see this change added upstream.

**Comment by acao** [comment]
I agree with @evocateur, this is what conventional commits is made for.
Another option that some projects may be able to use is scoping your `lerna version` and/or `lerna publish` commands:

`lerna version --scope @ourcourp/lib-* minor`
`lerna version --scope '{@ourcourp/*-service,@ourcourp/*-ui}' patch`

But really, if your versions are really important to whatever you're working on, and you arent following some kind of conventional commits standard, you are going to end up with a lot of complicated workarounds. What lerna has built in already for this is incredible. I just introduced a 300+ engineer team to it for our internal shared components and libraries (about 10 monorepos) and it's been well recieved.  You can customize the changelog preset, kick out releases on github or gitlab, handle pre-releases and graduation, etc. The configuration is incredibly minimal, and it ties together a lot of tooling that previously when our team researched this decision seemed a bit disparate, and in many cases just didnt exist yet for lerna. It was an exciting surprise when we noticed lerna 3 had all of these features built in.

Add commitlint or cz-cli on top and you have a nice quick multistep flow for each commit thats highly customizeable, and not really intrusive. Most of the popular IDEs have support for commitizen or commitlint via extensions, for those who are more git cli averse.

If you're averse to commits made by CI (this would make version bumps and changelog entries for each subpackage when using independent versions, for example), then developers can run `lerna version` standalone locally, and then CI can publish, or they can publish locally as well even, whatever makes the most sense for your workflow. You could have `package.json` scripts for `release`, `prerelease` and `graduate` (for bumping preprelease versions to full release)

**Comment by Perdiga** [comment]
We are in 2021 and I still want this :( ..  for me, the solution is using a set of yarn command to do this using Jenkins

**Comment by quannt** [comment]
Great thread with lots of useful information. Just want to chimp in and share that I have the same experience with @acao, we decided to adopt conventional commits, the setup was pretty straightforward and painless. 

Our flow looks like this
1. [Developer] Submit a pull request.
2. [Developer] Pull request is merged to the main branch.
3. [CI] Github Actions to run `lerna version --conventional-commits --yes --message 'chore: release'`
4. [CI] Github Actions to deploy to NPM.

**Comment by JamesHenry** [comment]
Hi Folks, for housekeeping I'm going to close this very old issue. I do agree with previous comments about this being out of scope for lerna, which has a well established relationship with conventional-commits.

If you want more nuanced and complex orchestration of releases (including across programming and language ecosystems) I recommend checking out the programmatic API of `nx release`: https://nx.dev/docs/guides/nx-release/programmatic-api - it shares the same internals as the CLI, but gives you full control of how and when the individual pieces are configured and invoked in custom scripts.

Many thanks!
